kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -42,9 +39,322 @@ import cats.syntax.either._
+          .getFailed
+          .asScala
+          .toList
+          .map { bree =>
+            (bree.getId, BatchResultErrorInfo(bree.getCode, bree.getMessage))
+          }
+          .toMap
+        // Events to retry and reasons for failure
+        msgGroup.collect {
+          case (e, m) if failures.contains(m.getId) =>
+            (e, failures(m.getId))
+        }
+      }
+    }
+
+  def toSqsMessages(events: List[Events]): List[(Events, SendMessageBatchRequestEntry)] =
+    events.map(e => (e, new SendMessageBatchRequestEntry(UUID.randomUUID.toString, b64Encode(e.payloads))))
+
+  def b64Encode(msg: Array[Byte]): String = {
+    val buffer = java.util.Base64.getEncoder.encode(msg)
+    new String(buffer)
+  }
+
+  def scheduleWrite(batch: List[Events], target: Target, lastBackoff: Long = minBackoff)(
+    retriesLeft: Int
+  ): Unit = {
+    val nextBackoff = getNextBackoff(lastBackoff)
+    executorService.schedule(
+      new Thread {
+        override def run(): Unit =
+          // If an SQS buffer is configured, keep switching targets every 10 retries.
+          // If no SQS buffer is configured, continue retrying to Kinesis forever.
+          if (retriesLeft > 0) sinkBatch(batch, target, nextBackoff)(retriesLeft)
+          else sinkBatch(batch, switchTarget(target), nextBackoff)(10)
+      },
+      lastBackoff,
+      MILLISECONDS
+    )
+    ()
+  }
+
+  // No-op if no SQS buffer is configured
+  def switchTarget(target: Target): Target = target match {
+    case KinesisStream(Some(sqs)) => SqsQueue(sqs)
+    case KinesisStream(None)      => KinesisStream(None)
+    case SqsQueue(sqs)            => KinesisStream(Some(sqs))
+  }
+
+  /**
+    * How long to wait before sending the next request
+    * @param lastBackoff The previous backoff time
+    * @return Minimum of maxBackoff and a random number between minBackoff and three times lastBackoff
+    */
+  private def getNextBackoff(lastBackoff: Long): Long =
+    (minBackoff + randomGenerator.nextDouble() * (lastBackoff * 3 - minBackoff)).toLong.min(maxBackoff)
+
+  def shutdown(): Unit = {
+    executorService.shutdown()
+    executorService.awaitTermination(10000, MILLISECONDS)
+    ()
+  }
+}
+
 /** KinesisSink companion object with factory method */
 object KinesisSink {
 
+  sealed trait Target
+  final case class KinesisStream(buffer: Option[SqsClientAndName]) extends Target
+  final case class SqsQueue(sqs: SqsClientAndName) extends Target
+
+  // Details about why messages failed to be written to SQS.
+  final case class BatchResultErrorInfo(code: String, message: String)
+
+  /**
+    * Events to be written to Kinesis or SQS.
+    * @param payloads Serialized events extracted from a CollectorPayload.
+    *                 The size of this collection is limited by MaxBytes.
+    *                 Not to be confused with a 'batch' events to sink.
+    * @param key Partition key for Kinesis
+    */
+  final case class Events(payloads: Array[Byte], key: String)
```

Comment:
Why is it plural ?

Reply:
Each `Events` object contains potentially multiple events. I found the singular confusing, because it created the impressions we're dealing with events one by one. But actually this represents a batch that was extracted from the CollectorPayload to match the Kinesis / SQS limits.